StepLang/Statements/StatementParser.cs

```diff
@@ -191,6 +194,70 @@ private async Task<Statement> ParseWhileLoop(Token whileToken, CancellationToken
         };
     }
 
+    private async Task<Statement> ParseForEachLoop(Token foreachToken, CancellationToken cancellationToken)
+    {
+        // looping: foreach ([[type name] <identifier>: ] [type name] <identifier> in <expression>) { [statement]* }
+
+        var openingParenthesesToken = tokenQueue.Dequeue(TokenType.OpeningParentheses);
+
+        var firstExpression = ParseForEachParamExpression();
+        Expression? keyVariableExpression;
+        Expression? valueVariableExpression;
+        if (tokenQueue.PeekType() is TokenType.ColonSymbol)
+        {
+            keyVariableExpression = firstExpression;
+
+            tokenQueue.Dequeue(TokenType.ColonSymbol);
+
+            valueVariableExpression = ParseForEachParamExpression();
+        }
+        else
+        {
+            keyVariableExpression = null;
+            valueVariableExpression = firstExpression;
+        }
+
+        tokenQueue.Dequeue(TokenType.InKeyword);
+
+        var iterableExpressionTokens = tokenQueue.DequeueUntil(TokenType.ClosingParentheses);
+        tokenQueue.Dequeue(TokenType.ClosingParentheses);
+
+        tokenQueue.Dequeue(TokenType.OpeningCurlyBracket);
+        var statementsTokens = tokenQueue.DequeueUntil(TokenType.ClosingCurlyBracket);
+        tokenQueue.Dequeue(TokenType.ClosingCurlyBracket);
+
+        var iterableExpression = await ExpressionParser.ParseAsync(iterableExpressionTokens, cancellationToken);
+        var statements = await ParseStatements(statementsTokens, cancellationToken).ToListAsync(cancellationToken);
+
+        return new ForEachStatement(keyVariableExpression, valueVariableExpression, iterableExpression, statements)
+        {
+            Location = foreachToken.Location,
+        };
+
+        Expression ParseForEachParamExpression()
+        {
+            if (tokenQueue.PeekType() is TokenType.TypeName)
+            {
+                var firstType = tokenQueue.Dequeue(TokenType.TypeName);
+                var firstIdentifier = tokenQueue.Dequeue(TokenType.Identifier);
+
+                return new VariableDeclarationExpression(firstType, firstIdentifier);
+            }
+
+            if (tokenQueue.PeekType() is TokenType.Identifier)
```

Comment:
would be more efficient to call `PeekType()` only once, no?

Reply:
It needs to be called in case `tokenQueue.Dequeue` has been executed before.

Reply:
OH. No you're right. Totally overlooked the `return` statement...